lib/bap_image/llvm_binary_stubs.hpp

```diff
@@ -0,0 +1,449 @@
+
+private:
+    std::string name_;
+    uint64_t addr_;
+    uint64_t size_;
+};
+
+std::vector<section> read(const ObjectFile* obj) {
+    int size = utils::distance(obj->begin_sections(),
+                               obj->end_sections());
+    std::vector<section> sections;
+    sections.reserve(size);
+    std::transform(obj->begin_sections(),
+                   obj->end_sections(),
+                   std::back_inserter(sections),
+                   [](const SectionRef& s) { return section(s); });
+    return sections;
+}
+
+} //namespace sec
+
+namespace img {
+using namespace llvm;
+using namespace llvm::object;
+
+struct image_base {
+    virtual uint64_t entry() const = 0;
+    virtual Triple::ArchType arch() const = 0;
+    virtual std::vector<seg::segment> segments() const = 0;
+    virtual std::vector<sym::symbol> symbols() const = 0;
+    virtual std::vector<sec::section> sections() const = 0;
+    virtual ~image_base() {}
+};
+
+Triple::ArchType image_arch(const ObjectFile* obj) {
+    return static_cast<Triple::ArchType>(obj->getArch());
+}
+
+template <typename ELFT>
+uint64_t image_entry(const ELFObjectFile<ELFT>* obj) {
+    return obj -> getELFFile() -> getHeader() -> e_entry;
+}
+
+uint64_t image_entry(const MachOObjectFile* obj) {  
+    typedef MachOObjectFile::LoadCommandInfo command_info;
+    typedef std::vector<command_info> commands;
+    typedef std::vector<command_info>::const_iterator const_iterator;
+    commands cmds = utils::load_commands(obj);
+    const_iterator it =
+        std::find_if(cmds.begin(), cmds.end(), 
+                     [](const command_info &info){
+                         return
+                         info.C.cmd == MachO::LoadCommandType::LC_MAIN;});
+    if (it != cmds.end()) {
+        const MachO::entry_point_command *entry_cmd = 
+            reinterpret_cast<const MachO::entry_point_command*>(it->Ptr);
+        return entry_cmd->entryoff;
+    } else {
+        llvm_binary_fail("LC_MAIN not found, binary version < 10.8");
+    }
+}
+
+uint64_t image_entry(const COFFObjectFile* obj) {
+    const pe32_header* hdr = 0;
+    if (error_code ec = obj -> getPE32Header(hdr))
+        llvm_binary_fail(ec);
+    if (!hdr)
+        llvm_binary_fail("PE header not found");
+    return hdr->AddressOfEntryPoint;
+};
+
+template <typename T>
+struct image : image_base {
+    explicit image(const T* obj) : obj_(obj) {}
+    Triple::ArchType arch() const {
+        return image_arch(obj_);
+    }
+
+    uint64_t entry() const {
+        return image_entry(obj_);
+    }
+
+    std::vector<seg::segment> segments() const {
+        return seg::read(obj_);
+    }
+
+    std::vector<sym::symbol> symbols() const {
+        return sym::read(obj_);
+    }
+
+    std::vector<sec::section> sections() const {
+        return sec::read(obj_);
+    }
+protected:
+    const T *obj_;
+};
+
+template <typename T>
+std::shared_ptr<image_base> create_image(const ObjectFile* obj) {
```

Comment:
it is not a create function, it is a cast function. Otherwise, calling `create_image` in each getter shocks the reader